crates/oxc_parser/src/modifiers.rs

```diff
@@ -231,55 +337,40 @@ impl<'a> ParserImpl<'a> {
     }
 
     fn modifier(kind: Kind, span: Span) -> Modifier {
-        let modifier_kind = match kind {
-            Kind::Abstract => ModifierKind::Abstract,
-            Kind::Declare => ModifierKind::Declare,
-            Kind::Private => ModifierKind::Private,
-            Kind::Protected => ModifierKind::Protected,
-            Kind::Public => ModifierKind::Public,
-            Kind::Static => ModifierKind::Static,
-            Kind::Readonly => ModifierKind::Readonly,
-            Kind::Override => ModifierKind::Override,
-            Kind::Async => ModifierKind::Async,
-            Kind::Const => ModifierKind::Const,
-            Kind::In => ModifierKind::In,
-            Kind::Out => ModifierKind::Out,
-            Kind::Export => ModifierKind::Export,
-            Kind::Default => ModifierKind::Default,
-            Kind::Accessor => ModifierKind::Accessor,
-            _ => unreachable!(),
-        };
-        Modifier { span, kind: modifier_kind }
+        Modifier { span, kind: kind.into() }
     }
 
     pub(crate) fn parse_modifiers(
         &mut self,
-        _allow_decorators: bool,
+        allow_decorators: bool,
         permit_const_as_modifier: bool,
         stop_on_start_of_class_static_block: bool,
     ) -> Modifiers<'a> {
         let mut has_seen_static_modifier = false;
-        // let mut has_leading_modifier = false;
-        // let mut has_trailing_decorator = false;
+        let mut has_leading_modifier = false;
+        let mut has_trailing_decorator = false;
+
         let mut modifiers = self.ast.new_vec();
+        let mut modifier_flags = ModifierFlags::empty();
 
         // parse leading decorators
-        // if (allowDecorators && token() === SyntaxKind.AtToken) {
-        // while (decorator = tryParseDecorator()) {
-        // list = append(list, decorator);
-        // }
-        // }
+        if allow_decorators && matches!(self.cur_kind(), Kind::At) {
+            self.try_parse(Self::eat_decorators);
+        }
 
         // parse leading modifiers
         while let Some(modifier) = self.try_parse_modifier(
             has_seen_static_modifier,
             permit_const_as_modifier,
             stop_on_start_of_class_static_block,
         ) {
-            if modifier.kind == ModifierKind::Static {
+            if modifier.is_static() {
                 has_seen_static_modifier = true;
             }
+            self.check_for_duplicate_modifiers(modifier_flags, &modifier);
+            modifier_flags.set(modifier.kind.into(), true);
             modifiers.push(modifier);
+            has_leading_modifier = true;
         }
 
         // parse trailing decorators, but only if we parsed any leading modifiers
```

Comment:
I wonder where does trailing decorators occur lol

Reply:
I think this is b/c of the syntax discrepancy between ts experimental decorators and
the ecmascript proposal?